usersession/xdgopenproxy/portal_launcher.go

```diff
@@ -133,36 +169,91 @@ func (p *portalLauncher) portalCall(bus *dbus.Conn, call func() (dbus.ObjectPath
 	}
 }
 
-func (p *portalLauncher) OpenFile(bus *dbus.Conn, filename string) error {
-	portal, err := p.desktopPortal(bus)
+func (p *portalLauncher) OpenFile(bus *dbus.Conn, filename string, options LauncherModifiers) error {
+
+	var attemptWritable bool
+	var fd int
+
+	pathStat, err := os.Stat(filename)
+	if err != nil {
+		return &responseError{msg: err.Error()}
+	}
+
+	// Trying to make paths writable is not valid, but doesn't fail here, so explicitly avoid bothering.
+	// Otherwise try open a file with RDRW and if that fails, try again with RDONLY.
+	if pathStat.IsDir() {
+		fd, err = syscall.Open(filename, syscall.O_RDONLY, 0)
+		if err != nil {
+			return &responseError{msg: err.Error()}
+		}
+	} else {
+		attemptWritable = true
+		fd, err = syscall.Open(filename, syscall.O_RDWR, 0)
+		if err != nil {
+			fd, err = syscall.Open(filename, syscall.O_RDONLY, 0)
+			if err != nil {
+				return &responseError{msg: err.Error()}
+			}
+			attemptWritable = false
+		}
+	}
+	defer syscall.Close(fd)
+
+	portal, err := p.documentsPortal(bus)
 	if err != nil {
 		return err
 	}
 
-	fd, err := syscall.Open(filename, syscall.O_RDONLY, 0)
+	// Try add the file to the documents portal, allowing OpenURI to mediate sandbox permissions if needed.
+	// Failure here is not a critical error, the application opened by OpenURI might not need mediation.
+	if attemptWritable {
+		var docID interface{}
+		err = portal.Call(documentsPortalDocumentsIface+".Add", 0, dbus.UnixFD(fd), false, false).Store(&docID)
+		if err != nil {
+			attemptWritable = false
+		}
+	}
+
+	portal, err = p.desktopPortal(bus)
 	if err != nil {
-		return &responseError{msg: err.Error()}
+		return err
 	}
-	defer syscall.Close(fd)
 
-	return p.portalCall(bus, func() (dbus.ObjectPath, error) {
+	return p.desktopPortalCall(bus, func() (dbus.ObjectPath, error) {
 		var (
 			parent  string
-			options map[string]dbus.Variant
 			request dbus.ObjectPath
 		)
-		err := portal.Call(desktopPortalOpenURIIface+".OpenFile", 0, parent, dbus.UnixFD(fd), options).Store(&request)
+
+		// The first call might fail if the file isn't actually writable by the host
+		// e.g, loading files in $SNAP due to the read only squashfs nature.
+		// Try with write enabled first and retry in the event of an error.
+		optionalFlags := make(map[string]dbus.Variant)
+		if options.Ask {
+			optionalFlags["ask"] = dbus.MakeVariant(true)
+		}
+		if attemptWritable {
+			optionalFlags["writable"] = dbus.MakeVariant(true)
+		}
+
+		err := portal.Call(desktopPortalOpenURIIface+".OpenFile", 0, parent, dbus.UnixFD(fd), optionalFlags).Store(&request)
+		// Only bother again if writable was set in the first place
+		if err != nil && attemptWritable {
```

Comment:
Is there a particular error we'd expect to get in this case? If so, it might make sense to check for that rather than retrying on all errors.

Reply:
The main error I'd noticed is that you can acquire a file descriptor with RDRW in the sandboxed environment for files in $SNAP. If you try open these with writable, the API gives an error that $SNAP is a read only filesystem. Aside from this, I'd imagine there might be other situations (e.g FUSE filesystems that might not accurately model Unix permissions?) that might have similar issues but I haven't seen any specifically.